Objects/moduleobject.c

```diff
@@ -819,68 +872,109 @@ _Py_module_getattro_impl(PyModuleObject *m, PyObject *name, int suppress)
         Py_DECREF(getattr);
         return result;
     }
+
+    if (suppress == 1) {
+        return NULL;
+    }
     if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__name__), &mod_name) < 0) {
         return NULL;
     }
-    if (mod_name && PyUnicode_Check(mod_name)) {
-        PyObject *spec;
-        if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+    if (!mod_name || !PyUnicode_Check(mod_name)) {
+        Py_XDECREF(mod_name);
+        PyErr_Format(PyExc_AttributeError,
+                    "module has no attribute '%U'", name);
+        return NULL;
+    }
+    PyObject *spec;
+    if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+        Py_DECREF(mod_name);
+        return NULL;
+    }
+    PyObject *origin = NULL;
+    if (spec) {
+        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
+        if (rc == -1) {
+            Py_DECREF(spec);
             Py_DECREF(mod_name);
             return NULL;
         }
-        if (suppress != 1) {
-            int rc = _PyModuleSpec_IsInitializing(spec);
+        if (rc == 1 && !PyUnicode_Check(origin)) {
+            Py_DECREF(origin);
+            origin = NULL;
+        }
+    }
```

Comment:
Would there be any value to falling back to `mod.__file__` for `origin` here?

```suggestion
    }
    if (origin == NULL) {
        int rc = PyObject_GetOptionalAttr(m, &_Py_ID(__file__), &origin);
        if (rc == -1) {
            Py_DECREF(spec);
            Py_DECREF(mod_name);
            return NULL;
        }
        if (rc == 1 && !PyUnicode_Check(origin)) {
            Py_DECREF(origin);
            origin = NULL;
        }
    }
```

Reply:
I think not, I feel in the cases this differs from `__spec__.origin` we'd be guessing too much. Let me know if you have a real world case in mind though!

Reply:
As a future reader, it might not be clear why `__file__` isn't factored in here.  Consider leaving a short note indicating that we could potentially fall back to `__file__` but currently there isn't any clear value in doing so.

Reply:
Added! :-)